pkg/maintenance/redis.go

```diff
@@ -0,0 +1,178 @@
+package maintenance
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"github.com/blang/semver/v4"
+	"github.com/crossplane-contrib/provider-helm/apis/release/v1beta1"
+	"github.com/go-logr/logr"
+	"github.com/spf13/viper"
+	"github.com/vshn/appcat/pkg/controller/postgres"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/types"
+	"net/http"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"strconv"
+)
+
+var redisURL = "https://hub.docker.com/v2/namespaces/library/repositories/redis/tags?page_size=100"
+var imageContentType = "image"
+var activeTagStatus = "active"
+
+// Result is the object that has details of a docker image tag
+type Result struct {
+	// Name is the tag
+	Name string `json:"name,omitempty"`
+	// TagStatus indicates if the tag is active
+	TagStatus string `json:"tag_status,omitempty"`
+	// ContentType indicates whether this Result is an image
+	ContentType string `json:"content_type,omitempty"`
+}
+
+// Payload contains the response body of a docker image tags
+type Payload struct {
+	Results []Result `json:"results,omitempty"`
+}
+
+type Redis struct {
+	K8sClient         client.Client
+	HttpClient        *http.Client
+	log               logr.Logger
+	instanceNamespace string
+}
+
+// DoMaintenance will run postgresql's maintenance script.
+func (r *Redis) DoMaintenance(ctx context.Context) error {
+	err := r.configure()
+	if err != nil {
+		return fmt.Errorf("cannot configure job to run: %v", err)
+	}
+
+	results, err := r.getVersions(redisURL)
+	if err != nil {
+		return fmt.Errorf("cannot get versions from Docker Hub: %v", err)
+	}
+
+	release, err := r.getRelease(ctx)
+	if err != nil {
+		return fmt.Errorf("cannot get release from namespace %s: %v", r.instanceNamespace, err)
+	}
+
+	version, isNewer, err := r.getNewVersion(release, results)
+	if err != nil {
+		return fmt.Errorf("cannot parse new version from release %s: %v", release.Name, err)
+	}
+	if !isNewer {
+		return nil
+	}
+
+	err = r.patchRelease(ctx, *version, release)
+	if err != nil {
+		return fmt.Errorf("cannot patch release %s: %v", release.Name, err)
+	}
+	return nil
+}
+
+func (r *Redis) patchRelease(ctx context.Context, version semver.Version, release *v1beta1.Release) error {
+	patchOps := []postgres.Jsonpatch{
+		{
+			Op:    "replace",
+			Path:  "/spec/forProvider/values/image/tag",
```

Comment:
This gets overwrite again by the P+T composition: https://github.com/vshn/component-appcat/blob/master/component/vshn_redis.jsonnet#L514

We'll have to handle this via comp-function.

Reply:
As agreed I added a new comp function that will update the helm values first.